### ISSUE TITLE
Allow to provide a user password for non interactive sessions

### ISSUE THREAD

**Issue by v1tzl1** [esnet/iperf#706]
# Enhancement Request

* Current behavior
When using authentication the iperf3 client will ask for the password on the command line. While this is very useful for manual testing it makes automated testing using authentication (e.g. in a cronjob) very hard.

* Desired behavior
Give an option to specify the password. Either as a new command line option, or attempt to read it from a (set of) hardcoded file(s) first and only ask for it once non was found.

I'd be happy to provide an implementation of the second approach.

**Comment by bmah888** [comment]
Seems like a reasonable thing to want to do.  I have a preference for a command-line option...I think it'd be easier than the second approach.  Also for all UNIX utilities I can think of that allow non-interactive input of passwords, they all just support a command-line option.

**Comment by ralcini** [comment]
@bmah888 command line option for a password? So the password of a user remain in the bash history? Or can be seen with a "ps aux" command? 

I don't think it's a good solution. Why not a environmental variable? ..or read password from a file?

**Comment by bmah888** [comment]
@ralcini Fair enough, although I had in mind the precedent set by the Postgres and MySQL command-line clients.  Admittedly part of my dislike for the second option in the original comment was the "hard-coded" pathname part.

**Comment by v1tzl1** [comment]
@ralcini Good spot!

Indeed when suggesting a command line option I was thinking of the mysql client. Although their man page specifically states that providing a password as option should be considered insecure. Specifying the -p option without anything else will make mysql ask for the password on the command line, which iperf3 already does without any command line option.

I suggested the hardcoded files, because I already spotted the routine where the password is read in, so writing a patch to do that would be fairly easy and local, while a command line parameter would be needed to be passed through the call tree. (Which might be easy as well, I did not look at the entire iperf3 code yet).

So I suppose the best option is to add a command line option to specify a file in which the password is stored.

**Comment by ralcini** [comment]
In my personal opinion: I agree with @v1tzl1 for a file containing the password.. the default path of the file can be the user home but it must be overridable with a command line option..

**Comment by bmah888** [comment]
OK.  You've got me convinced on using a file to hold the password.  But iperf3 should really default to asking the user.  If iperf3 can't find that file (either because the file doesn't exist, can't be read, etc.) it should error out.  For some reason I'm not a big fan of having a default location for the file, although there are plenty of other programs that have defaults.  I guess I'm OK either way on that.

**Comment by TheRealDJ** [comment]
How about an environment variable? History will show the variable name, not the value.

# export PASSWORD=insecure
# echo $PASSWORD
 insecure
# history
 ...
 echo $PASSWORD
 ...

**Comment by ralcini** [comment]
@TheRealDJ I agree with you, but the insecure way to use this method by the users is:

`PASSWORD=mypwd iperf3 .....`

Other aspect to verify is that iperf in debug mode do not logs all the content of enviroment...

**Comment by pjebs** [comment]
I prefer command line as an option. I agree with @ralcini that it is dangerous and gets saved in the history -- but that is not an issue for me. I am using short-lived docker containers. I think the option should be given with a clear warning stated in the documentation.
